parse_xml.py

```diff
@@ -21,115 +22,207 @@ def changeExt(fname, ext):
                 else:
-                    logging.info("TXT file: "+docid+".txt  does not exist. JSON file creation skipped.")
-                    return 
-
-            #transform output to json and save to file with same name
-            with open(fn,'w') as outfile:
-                json.dump(doc,outfile)
-                logging.info("-- Processing of XML file complete")
-                #process new JSON file
+                    for x in doc['main']['DATA_RECORD']:
+                        docid = x["DOCUMENT_IMAGE_ID"]
+                        txtfn = os.path.join(os.path.dirname(fn),'PDF_image',docid+'.txt')
+                        if os.path.isfile(txtfn):
+                            kitems = ["DOCUMENT_CREATE_DT","LAST_MODIFIED_TS","PATENT_ISSUE_DT","DECISION_MAILED_DT","PRE_GRANT_PUBLICATION_DT","APPLICANT_PUB_AUTHORIZATION_DT"]
+                            for item in kitems:
+                                x[item] = formatDate(x.pop(item))
+                            kitems2 = {"appid":"BD_PATENT_APPLICATION_NO","doc_date":"DOCUMENT_CREATE_DT"}
+                            for key,value in kitems2.items():
+                                replaceKey(x,key,value)
+                            x['textdata'] = readDoc(txtfn)
+                        else:
+                            logging.error("File: "+txtfn+" does not exist.  Parsing of XML will be skipped")
+            #transform output to json
+            outputFile(fn,json.dumps(doc))
+            logging.info("-- Processing of XML file complete")
+        else:
+            logging.info("-- File: "+fn+" already exists.")
+
+    except KeyError as e:
+        logging.error("-- File: "+fn+" Key Error: {0} ".format(e))
+        pass
     except IOError as e:
-        logging.error("I/O error({0}): {1}".format(e.errno,e.strerror))
+        logging.error("-- I/O error({0}): {1}".format(e.errno,e.strerror))
     except:
-        logging.error("Unexpected error:", sys.exc_info()[0])
+        logging.error("-- Unexpected error:", sys.exc_info()[0])
         raise
 
-#validate date
-def validDate(s):
+def readJSON(fname):
     try:
-        datetime.strptime(s, "%Y%m%d")
-        return s
-    except ValueError:
-        msg = "Not a valid date: '{0}'.".format(s)
-        raise argparse.ArgumentTypeError(msg)
+        with open(fname) as fd:
+            doc = json.loads(fd.read())
+            if args.ftype in ("g","p"):
+                element = "us-patent-"+filetype[3]
+                docvalue = ["us-bibliographic-data-"+filetype[3],"publication-reference","document-id","appid"]
+                logfn = fname.split("_alt.json",1)[0]+"_solrlog.txt"
+            else:
+                element = "DATA_RECORD"
+                docvalue = ["DOCUMENT_IMAGE_ID"]
+                logfn = os.path.splitext(fname)[0]+"_solrlog.txt"
+            for x in doc['main'][element]:
+                x2 = x
+                for val in docvalue:
+                    x2 = x2[val]
```

Comment:
How does this work? It seems like you could run into errors. `x2` is being reassigned a scalar value in each loop, yet you're assigning it _from_ `x2` on each go? Seems like only the first hit would work.